cmd/e2e-test/main.go

```diff
@@ -0,0 +1,286 @@
+package main
+
+import (
+	"flag"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"go.uber.org/zap"
+)
+
+type TestSchema struct {
+	Input      []string     `toml:"input"`           // carbon-clickhouse input
+	ConfigTpl  string       `toml:"config_template"` // carbon-clickhouse config template
+	Clickhouse []Clickhouse `yaml:"clickhouse"`
+
+	Verify []Verify `yaml:"verify"`
+}
+
+type MainConfig struct {
+	Test *TestSchema `toml:"test"`
+}
+
+func IsDir(filename string) (bool, error) {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return info.IsDir(), nil
+}
+
+func expandDir(dirname string, paths *[]string) error {
+	files, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			if err = expandDir(path.Join(dirname, file.Name()), paths); err != nil {
+				return err
+			}
+		} else {
+			ext := path.Ext(file.Name())
+			if ext == ".toml" {
+				*paths = append(*paths, path.Join(dirname, file.Name()))
+			}
+		}
+	}
+
+	return nil
+}
+
+func expand(filename string, paths *[]string) error {
+	if len(filename) == 0 {
+		return nil
+	}
+	isDir, err := IsDir(filename)
+	if err == nil {
+		if isDir {
+			if err = expandDir(filename, paths); err != nil {
+				return err
+			}
+		} else {
+			*paths = append(*paths, filename)
+		}
+	}
+	return err
+}
+
+func main() {
+	_, filename, _, _ := runtime.Caller(0)
+	rootDir := path.Dir(path.Dir(path.Dir(filename))) // carbon-clickhouse repositiry root dir
+
+	config := flag.String("config", "", "toml configuration file or dir where toml files is searched (recursieve)")
+	verbose := flag.Bool("verbose", false, "verbose")
+	flag.Parse()
+	logger, err := zap.NewProduction()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var configs []string
+	err = expand(*config, &configs)
+	if err != nil {
+		logger.Fatal(
+			"config",
+			zap.Error(err),
+		)
+	}
+	if len(configs) == 0 {
+		logger.Fatal("config should be non-null")
+	}
+
+	failed := 0
+	total := 0
+	for _, config := range configs {
```

Comment:
It's quite difficult to follow the logic in the loop. Maybe, create a function for config processing?